### PR TITLE
better Point type factoring

### PR DESCRIPTION
**Pull request by avieth**
There is now a type for identifying actual blocks, called BlockPoint.
It's parameterized on the slot and hash, and specialized to use the
class/families variant, so that it's not forced upon the programmer.

Next, there is a Point functor which is just like Maybe; it adds the
Origin.

This way, we eliminate silly things like `Point 42 GenesisHash`. The
"genesis point" is now called Origin, and it has no slot number nor
header hash.

According to `Ouroboros.Network.Chain`, `headSlot Genesis = SlotNo 0`.
Also, valid chains have strictly increasing slot numbers. So in this
model, the first actual block has `SlotNo 1`. But that's contrary to the
`cardano-sl` implementation, which gives `SlotNo 0` to the first block.
In `ouroboros-consensus`, does any block have `SlotNo 0`?
`Point 0 (BlockHash hash)` is apparently nonsense too, for all `hash`.

Motivation for this change: I'm reworking the hash index in byron-proxy
so that it will be ready for use with the `ChainDB` (currently it is paired
with an `ImmutableDB`). This means the index must deal with forks.
The plan is to have it follow by way of the `newReader` interface. Naturally,
it will use `Point`s to deal with rollbacks. *I don't want to (nor should I have to)
write a program which knows how to rollback to the point at slot 42 with the
genesis hash. This makes no sense. With this patch, I won't have to.*

The patch only goes so far as to make `ouroboros-network` compile and its tests
pass. I will patch downstream packages only if this pull request is approved in
principle.

### CODE REVIEW

**Comment by avieth** [comment]
There is

```Haskell
rewindChainState :: OuroborosTag p => NodeConfig p -> ChainState p -> SlotNo -> Maybe (ChainState p)
```

What does this mean at `SlotNo 0`? Does it remove all blocks? Or does the block at `SlotNo 0` remain? Is there ever a block at `SlotNo 0`? In mainnet, there is, but in the `ouroboros-network` model, head of the empty chain `Genesis :: Chain` has slot `SlotNo 0`.

**Comment by coot** [comment]
@avieth it perfectly makes sense to have
```
data Chain block = Genesis | Chain block :> block
```
and and the same time:
```
data Point block = Origin | Point block
```
But let's see what the consensus team things about it, @mrBliss, this might also affect `cardano-ledger`
integration with `ouroboros-consensus`.

File: ouroboros-network/src/Ouroboros/Network/Block.hs [hunk 1]
```diff
@@ -108,27 +116,37 @@ castHash (BlockHash b) = BlockHash b
...
+data TBlockPoint slot hash = TBlockPoint
+  { pointSlot :: slot
+  , pointHash :: hash
+  }
+  deriving (Eq, Ord, Show)
+
+type BlockPoint block = TBlockPoint SlotNo (HeaderHash block)
+
+-- | `BlockPoint` with an origin point.
+data TPoint t where
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Block.hs]: Hmm, why not call this type just `Point`?  I also think that using `block` instead of `t` is better, for the consistency sake.  A comment can explain that this can be a block or header, and a comment could explain that `Origin :: TPoing block` corresponds to `Genesis :: Chain block`.

File: ouroboros-network/src/Ouroboros/Network/Block.hs [hunk 2]
```diff
@@ -139,12 +157,14 @@ blockPoint b =
...
-  deriving (Eq, Show)
+
+deriving instance (Show block, Show (HeaderHash block)) => Show (ChainUpdate block)
+deriving instance (Eq block, Eq (HeaderHash block)) => Eq (ChainUpdate block)
 
 mapChainUpdate :: HeaderHash a ~ HeaderHash b
                => (a -> b) -> ChainUpdate a -> ChainUpdate b
 mapChainUpdate f (AddBlock b) = AddBlock $ f b
-mapChainUpdate _ (RollBack p) = RollBack $ castPoint p
+mapChainUpdate _ (RollBack p) = RollBack $ p
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Block.hs]: the  `$` is redundant here, though I guess you left if to align things, so it's fine.

File: ouroboros-network/src/Ouroboros/Network/Chain.hs [hunk 3]
```diff
@@ -289,22 +293,23 @@ findFirstPoint
   :: HasHeader block
   => [Point block]
   -> Chain block
-  -> Maybe (Point block)
-findFirstPoint [] _     = Nothing
+  -> Point block
+findFirstPoint [] _     = Origin
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Chain.hs]: This is where it becomes strange since now `Origin` means two things: not found and if `Origin` was included in the `ps :: [Point block]` list it may return `Origin` because it's found. I'd argue that this is not important, because including `Origin` makes not much sense in the first place.

Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Chain.hs]: `Origin` is on every `Chain block` so yes, `findFirstPoint (Origin : ps) chain = findFirstPoint ps chain`. It's not so strange. There's a similar story for `False` and `or :: [Bool] -> Bool`.

File: ouroboros-network/src/Ouroboros/Network/Block.hs (hunk 1 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Block.hs]: Would be better as `Point` but I used a new name so that I wouldn't have to make as many name changes in downstream modules.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Block.hs]: I see, we can claim that `T` stands for `ThePoint`.

File: ouroboros-network/src/Ouroboros/Network/Chain.hs (hunk 3 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Chain.hs]: On second thought, it actually is important to distinguish the case when none of the given points are on the chain. Comparison to `or` isn't good, since `findFirstPoint` is primarily about membership in the `[Point block]` list. I'll change it back to give `Maybe (Point block)`

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Chain.hs]: This makes sense.

**Comment by avieth** [comment]
Pushed another commit which makes ouroboros-consensus compile and tests pass.

**Comment by avieth** [comment]
One thing I noticed is that `Ouroboros.Network.Chain.validExtension` used to rule out any block that had `SlotNo 0`. While that is in principle a legitimate way to go, I think it's not ideal. It reserves `SlotNo 0` for something which does not correspond to a block (the genesis is not a block). In the new factoring introduced in this patch, every `SlotNo n` is as good as any other when it comes to indexing a chain, with no special treatment for `n = 0`. Then, if the programmer wishes to include the genesis, they can throw in `Origin` by way of `TPoint`.

A nice summary of the benefit of this patch can be seen in `Ouroboros.Storage.ChainDB.Model`, where the partial function `notGenesis :: Point blk -> HeaderHash blk` has been removed.

**Comment by avieth** [comment]
@dcoutts @coot checks passed

**Comment by avieth** [comment]
@mrBliss This `Point` type change will affect a `ChainDB` implementation so it would be good to hear your thoughts on it.

**Comment by mrBliss** [comment]
> @mrBliss This `Point` type change will affect a `ChainDB` implementation so it would be good to hear your thoughts on it.

Ok, I'll give this a look after the weekend.

File: ouroboros-network/src/Ouroboros/Network/Block.hs [hunk 4]
```diff
@@ -108,27 +116,37 @@ castHash (BlockHash b) = BlockHash b
...
-     }
-   deriving (Eq, Ord, Show)
+-- | A block is identified by a slot and a hash.
+data TBlockPoint slot hash = TBlockPoint
+  { pointSlot :: slot
+  , pointHash :: hash
+  }
+  deriving (Eq, Ord, Show)
+
+type BlockPoint block = TBlockPoint SlotNo (HeaderHash block)
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Block.hs]: I'm a bit grumpy about having to have two different names here. I don't see why we need the T variant with its extra params. More names makes the API bigger and harder to remember and harder for new people to get into.

The `HeaderHash` is already a standalone type family, so surely it's already flexible enough. And it's not like there's many choices for the representation of `SlotNo`.

File: ouroboros-network/src/Ouroboros/Network/Block.hs (hunk 1 above)
Comment (dcoutts) [reply ouroboros-network/src/Ouroboros/Network/Block.hs]: I think we all agree that distinguishing the genesis point from other block points is a good thing, lets do it fully, even if it means more downstream changes. Fewer names in the API is a good thing.

File: ouroboros-network/src/Ouroboros/Network/Block.hs [hunk 5]
```diff
@@ -108,27 +116,37 @@ castHash (BlockHash b) = BlockHash b
...
+  fmap _ Origin    = Origin
+  fmap f (Point t) = Point (f t)
+
+fromTPoint :: r -> (t -> r) -> TPoint t -> r
+fromTPoint r _ Origin    = r
+fromTPoint _ f (Point t) = f t
 
-castPoint :: (HeaderHash a ~ HeaderHash b) => Point a -> Point b
-castPoint (Point a b) = Point a (castHash b)
+type Point block = TPoint (BlockPoint block)
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Block.hs]: But why do we need the two variants of `Point`? Why not simply:
```
data Point block = GenesisPoint | Point (BlockPoint block)
```
Is it really that much of a benefit to it being a functor? Can we handle those (presumably rather rare) cases by conversion to/from `Maybe (BlockPoint block)`?

Fewer names! Fewer concepts :-)

File: ouroboros-network/src/Ouroboros/Network/Block.hs [hunk 6]
```diff
@@ -139,12 +157,14 @@ blockPoint b =
 --
 data ChainUpdate block = AddBlock block
                        | RollBack (Point block)
-  deriving (Eq, Show)
+
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Block.hs]: As an aside, the consensus package has a chain update type like:
```
data ChainUpdate blk a =
    AddBlock a
  | RollBack (Point blk)
  deriving (Show, Functor)
```
This allows the point type and block type to be independent. This is useful when you have headers or blocks, but they both share the same point type.

We might want to move that version into the network package.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs [hunk 7]
```diff
@@ -2,6 +2,8 @@
 {-# LANGUAGE ScopedTypeVariables        #-}
 {-# LANGUAGE TypeFamilies               #-}
 {-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE AllowAmbiguousTypes        #-}
+{-# LANGUAGE TypeApplications           #-}
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs]: Ugg. Do we have to?

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs [hunk 8]
```diff
@@ -157,7 +161,7 @@ fetchDecisions fetchDecisionPolicy@FetchDecisionPolicy {
   . map swizzleI
 
     -- Filter to keep blocks that have not already been downloaded.
-  . filterNotAlreadyFetched
+  . filterNotAlreadyFetched @header @block
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs]: Presumably we would not need this if we were not so general with `TPoint`.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/State.hs [hunk 9]
```diff
@@ -258,7 +264,8 @@ data FetchStateFingerprint peer header block =
        !(Maybe (Point block))
        !(Map peer (Point header))
        !(Map peer (PeerFetchStatus header))
-  deriving Eq
+
+deriving instance (Ord peer, Eq header, Eq (HeaderHash header), Eq (HeaderHash block)) => Eq (FetchStateFingerprint peer header block)
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/State.hs]: This seems quite suspicious. Why are we now comparing headers? Surely we only need to compare points.

File: ouroboros-network/src/Ouroboros/Network/ChainProducerState.hs [hunk 10]
```diff
@@ -248,11 +256,14 @@ rollback p (ChainProducerState c rs) =
...
-    rs' = [ if pointSlot p' > pointSlot p
+    rs' = [ if betterPoint p' p
               then r { readerPoint = p, readerNext = ReaderBackTo }
               else r
           | r@ReaderState { readerPoint = p' } <- rs ]
-
+    betterPoint Chain.Origin _ = False
+    betterPoint _ Chain.Origin = True
+    betterPoint (Chain.Point p') (Chain.Point p'') =
+      Chain.pointSlot p' > Chain.pointSlot p''
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/ChainProducerState.hs]: Why do we need this? We have an `Ord` instance for `Point`.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/Examples.hs [hunk 11]
```diff
@@ -127,31 +128,30 @@ recentOffsets = [0,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584]
...
-                          ( HasHeader header
+chainSyncServerExample :: forall header m a.
+                          ( HasHeader header -- FIXME misleading language "Has"/"Is"
                           , MonadSTM m
-                          , HeaderHash header ~ HeaderHash blk
                           )
                        => a
                        -> TVar m (ChainProducerState header)
-                       -> ChainSyncServer header (Point blk) m a
+                       -> ChainSyncServer header (Point header) m a
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/Examples.hs]: This looks like a reasonable change, but why is it in this patch? It looks unrelated.

File: ouroboros-network/test/Test/AnchoredFragment.hs [hunk 12]
```diff
@@ -162,37 +162,37 @@ prop_addBlock (TestAddBlock c b) =
...
-  AF.withinFragmentBounds p c ==>
-  case AF.successorBlock p c of
-    Nothing -> AF.headPoint c === p
-    Just b  -> property $ AF.withinFragmentBounds (AF.blockPoint b) c
-          .&&. blockPrevHash b === pointHash p
+  AF.withinFragmentBounds (Point p) c ==>
+  case AF.successorBlock (Point p) c of
+    Nothing -> AF.headPoint c === Point p
+    Just b  -> property $ AF.withinFragmentBounds (Point (AF.blockPoint b)) c
+          .&&. blockPrevHash b === BlockHash (pointHash p)
```
Comment (dcoutts) [inline ouroboros-network/test/Test/AnchoredFragment.hs]: Lots and lots of BlockPoint -> Point conversion going on. I wonder if we couldn't pick some generators or utils so that we don't need quite so many.

**Comment by dcoutts** [comment]
Alternative go at starting this in #690. See commentary there.

**Comment by mrBliss** [comment]
I wholeheartedly agree with Duncan.

I'm all for refactoring  the `Point` type to rule out things like `Point 42 GenesisHash`.

I also understand why you want to assign `SlotNo 0` to Genesis, especially since `cardano-sl` is using `SlotNo 0` for a regular block. However, this will be a less straight-forward change than the `Point` refactoring, so I'd rather do this in a separate PR.

We're also using `BlockNo 0` for Genesis, would that have to change too?

Just a thought: we need a separate case for Genesis in multiple types: `Point`, `ChainHash`, and now for `SlotNo`. Maybe we can use one sum type, isomorphic to `Maybe`, for this (names are still temporary):
```haskell
data BlockOrGen a = Genesis | Block a
  deriving (Functor)

data BlockPoint blk = Point SlotNo (HeaderHash blk)
type Point      blk = BlockOrGen (BlockPoint blk)
type GenSlotNo      = BlockOrGen SlotNo
type ChainHash  blk = BlockOrGen (HeaderHash blk)
```

File: ouroboros-network/src/Ouroboros/Network/Block.hs (hunk 5 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/Block.hs]: I will make a change to get rid of the `T`-prefix. It's there because I wanted less of a fallout from a name change.

> Is it really that much of a benefit to it being a functor?

Yes. It's used in `Ouroboros.Network.Chain.{headSlot, headHash}` rather effectively. Again in `Ouroboros.Network.AnchoredFragment.{headSlot, lastSlot}`.

> Fewer names! Fewer concepts :-)

In my experience this is not always better. It can in fact make things worse. Why have nominal types at all?

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs (hunk 7 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs]: I suppose this is because `BlockPoint block` does not contain a `block`, only a `HeaderHash block`, and `HeaderHash` is not injective. Previously, `fetchDecisions` had a `(Point block -> Bool)` parameter, but now it's `(BlockPoint block -> Bool)`, which makes more sense, because obviously you have the origin point. Why did it work with the old `Point block` though? That one has a `ChainHash block` which also does not actually determine the `block` type, only its image under `HeaderHash`.

**Comment by avieth** [comment]
> I also understand why you want to assign SlotNo 0 to Genesis, especially since cardano-sl is using SlotNo 0 for a regular block. However, this will be a less straight-forward change than the Point refactoring, so I'd rather do this in a separate PR.
>
> We're also using BlockNo 0 for Genesis, would that have to change too?

The first slot is slot 0. If there is a block here, its block number is 1.

> Just a thought: we need a separate case for Genesis in multiple types: Point, ChainHash, and now for SlotNo. Maybe we can use one sum type, isomorphic to Maybe, for this (names are still temporary):
> 
> ```Haskell
> data BlockOrGen a = Genesis | Block a
>   deriving (Functor)
> 
> data BlockPoint blk = Point SlotNo (HeaderHash blk)
> type Point      blk = BlockOrGen (BlockPoint blk)
> type GenSlotNo      = BlockOrGen SlotNo
> type ChainHash  blk = BlockOrGen (HeaderHash blk)
```

This is exactly what I've done... but I've called `BlockOrGen` `TPoint` instead and everybody seems to hate the name. No problem, I always intended to change it.

**Comment by mrBliss** [comment]
> > Just a thought: we need a separate case for Genesis in multiple types: Point, ChainHash, and now for SlotNo. Maybe we can use one sum type, isomorphic to Maybe, for this (names are still temporary):
> > ```haskell
> > data BlockOrGen a = Genesis | Block a
> >   deriving (Functor)
> > 
> > data BlockPoint blk = Point SlotNo (HeaderHash blk)
> > type Point      blk = BlockOrGen (BlockPoint blk)
> > type GenSlotNo      = BlockOrGen SlotNo
> > type ChainHash  blk = BlockOrGen (HeaderHash blk)
> > ```
> 
> This is exactly what I've done... but I've called `BlockOrGen` `TPoint` instead and everybody seems to hate the name. No problem, I always intended to change it.

I see. Yeah, the name `TPoint` confused me, it made me think it was only meant for points, not for slots or hashes.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs (hunk 7 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs]: Alternatively, I could use `Proxy`s to get rid of the ambiguity. Would you prefer that?

File: ouroboros-network/src/Ouroboros/Network/ChainProducerState.hs (hunk 10 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/ChainProducerState.hs]: That `Ord` instance is derived, so it compares on the hash as well. For points at equal slots with different hashes, one would be determined to be better according to the hash ordering, which is weird and wrong. With this implementation `betterPoint`, the new point is discarded if its slot is not strictly greater.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs (hunk 7 above)
Comment (avieth) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs]: >  Why did it work with the old Point block though? That one has a ChainHash block which also does not actually determine the block type, only its image under HeaderHash.

I'm guessing because wrapping the `HeaderHash block` in a `newtype` or `data` with the `block` as a parameter makes in unambiguous.

**Comment by avieth** [comment]
Obsoleted by #706
